readhomer_atlas/library/named_entities.py

```diff
@@ -0,0 +1,34 @@
+import csv
+
+from .models import NamedEntity, Node
+
+
+def apply_named_entities(force=True):
+    if force:
+        NamedEntity.objects.all().delete()
+
+    named_entites_path = "data/annotations/named-entities/raw/named_entities.csv"
+    lookup = {}
+    with open(named_entites_path) as f:
+        reader = csv.DictReader(f)
+        for row in reader:
+            named_entity, _ = NamedEntity.objects.get_or_create(
+                urn=row["urn"],
+                defaults={
+                    "title": row["label"],
+                    "description": row["description"],
+                    "url": row["link"],
+                },
+            )
+            lookup[named_entity.urn] = named_entity
+
+    iliad_path = "data/annotations/named-entities/raw/tlg0012.tlg001.perseus-grc2.csv"
+    with open(iliad_path) as f:
+        reader = csv.DictReader(f)
+        for row in reader:
+            named_entity = lookup[row["named_entity_urn"]]
+            text_part = Node.objects.get(urn=row["ref"])
+            # @@@ resolve idx / position mismatch
+            position = int(row["token_position"]) + 1
```

Comment:
@jtauber something I wanted to discuss with you, from both a "data" standpoint and a "modeling" standpoint.

In the data you've provided in the CEX, the offsets are zero-based (and that's totally fine because I think that's what was provided in [sample.cex](https://trello-attachments.s3.amazonaws.com/5e46dff3ff311c824532b4b2/5e68e449ee410661d151237b/b8c329950cdc7fd9ba6d41fb27e9db20/sample.cex) on the Trello card.

But looking at [morphgnt/sblgnt](https://github.com/morphgnt/sblgnt/), I see that your identifiers were one-based:

```
010101 N- ----NSF- Βίβλος Βίβλος βίβλος βίβλος
010101 N- ----GSF- γενέσεως γενέσεως γενέσεως γένεσις
...
010102 N- ----NSM- Ἀβραὰμ Ἀβραὰμ Ἀβραάμ Ἀβραάμ
010102 V- 3AAI-S-- ἐγέννησεν ἐγέννησεν ἐγέννησε(ν) γεννάω
```

and the same within [Open Epictetus](https://jtauber.github.io/enchiridion/sentences/001.html):

```
01.01.001
01.01.002
```
I think that our token exemplars _ought to be_ 1-based, and if you agree I'm happy to make the changes to the CEX and we establish that going forward.

Reply:
On the modeling side, within ATLAS I've been using an `idx` field that has ended up being zero-based; the idx is relative to all of the tokens in a version, etc.  I kept it as zero-based because it was inherited from some of the slicing we were doing with the original text server.

For our `Token` model, we have a one-based `position`, where the value is always relative to the parent text part; this [sample query](https://explorehomer-feature-na-olfkfu.herokuapp.com/graphql/#query=%7B%0A%20%20tokens(first%3A%2010)%20%7B%0A%20%20%20%20edges%20%7B%0A%20%20%20%20%20%20node%20%7B%0A%20%20%20%20%20%20%20%20idx%0A%20%20%20%20%20%20%20%20position%0A%20%20%20%20%20%20%20%20textPart%20%7B%0A%20%20%20%20%20%20%20%20%20%20urn%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%7D%0A%20%20%7D%0A%7D%0A&variables=null&operationName=undefined) helps to demonstrate.

I think that the `idx` remains an internal thing to ATLAS, whereas the "position" is something that could be an external ID for standoff annotation (and that's actually what we've been doing in Digital Sira to date).

Does that seem reasonable?

Reply:
Note that MorphGNT as quoted doesn't have token IDs at all, only book-chapter-verse.

But yes, Open Epictetus is 1-indexed tokens in terms of the citation.

I agree with you that the citable references should be 1-indexed even if we use 0-indexing internally in some other places.

So yes, the CEX for the named entities should be changed to be 1-indexed.

Reply:
(sorry, I'll let you make this resolved)

Reply:
Thanks for weighing in, @jtauber!